pgwire: rework DecodeOidDatum to DecodeDatum to parse OidFamily types

otan opened the pull request:
Resolves #56193 

pgwire: rework DecodeOidDatum to DecodeDatum to parse OidFamily types

Reworked DecodeOidDatum to DecodeDatum to take in a type, which
encodes additional useful information necessary for ENUMs and oid
family types.

Release note (bug fix): Fixed a bug where reg* types were not parsed
properly over pgwire, COPY or prepared statements.

tree: create ParseDOid method

* Move ParseDOid and associated methods to new function.
* Move datum_test.go to datum_integration_test.go, as it does not import
  datum.go tests.
* Move some of datum_invariants_test.go out into datum_test.go.
* Created new datum_test.go with pure unit tests.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/56298)
<!-- Reviewable:end -->

otan committed:
tree: create ParseDOid method

* Move ParseDOid and associated methods to new function.
* Move datum_test.go to datum_integration_test.go, as it does not import
  datum.go tests.
* Move some of datum_invariants_test.go out into datum_test.go.
* Created new datum_test.go with pure unit tests.

Release note: None

rafiss reviewed (commented):
lgtm after fixing the comment. nice refactor ^_^

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56298#-:-MMHzP_d8VX624xMQ52L:b-r6ybi0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/pgwire/pgwirebase/encoding.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56298#-MMHx5Tp-naIz1YCiqon:-MMHx5Tp-naIz1YCiqoo:bjvukjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b38b9dbe3201a0c6204cbb92dea966262de6b9/pkg/sql/pgwire/pgwirebase/encoding.go#L300)):*
> ```
> If res is nil
> ```
nit: need to update comment since res is gone


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/sem/tree/casts_test.go:
```diff
@@ -203,77 +203,3 @@ func TestTupleCastVolatility(t *testing.T) {
 	}
 }
-
-func TestCastStringToRegClassTableName(t *testing.T) {
```
oh just noticed, where did this test go?

otan replied:
pkg/sql/sem/tree/datum_test.go

otan committed:
pgwire: rework DecodeOidDatum to DecodeDatum to parse OidFamily types

Reworked DecodeOidDatum to DecodeDatum to take in a type, which
encodes additional useful information necessary for ENUMs and oid
family types.

Release note (bug fix): Fixed a bug where reg* types were not parsed
properly over pgwire, COPY or prepared statements.

otan commented:
bors r=rafiss
